### ISSUE TITLE
Query: fix edge cases around Cast/OfType result operator lifting

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#8334]
There are some issues around result operator lifting, specifically Cast and OfType:

1.) cast used as type validation doesn't always throw when it should, e.g.:

```
from g in ctx.Gears.Cast<Officer>().Cast<Gear>()
select g.Nickname
```

when run on the client it will throw, because not every gear is an officer. However it works fine on sql sever, even though during QM optimization we turn this to:

```
from g in ctx.Gears
select ((Gear)((Officer)g)).Nickname
```

2.) Before we try to lift cast, we handle OfType, by updating the type on EntityQueryable directly, e.g:

```
from o in (from g in EntityQueryable<Gear> select g).OfType<Officer>()
select o.NumberOfReports
```

will get converted to:
```
from o in EntityQueryable<Officer>
select o.NumberOfReports
```

This however breaks down if Cast is introduced to the mix:

```
from o in (from g in EntityQueryable<Gear> select g).Cast<Gear>().OfType<Officer>()
select o.NumberOfReports
```

will fail, because we process OfType first:

```
from o in (from g in EntityQueryable<Officer> select g).Cast<Gear>()
select o.NumberOfReports
```

which then gets lifted to:
```
from o in (from g in EntityQueryable<Officer> select g)
select ((Gear)o).NumberOfReports
```

which is wrong

**Comment by maumar** [comment]
2.) can be addressed by removing all Cast operators that come before the given OfType, but that potentially makes 1) worse

**Comment by ajcvickers** [comment]
@maumar Can you write down some of the ideas we discussed in triage?

**Comment by maumar** [comment]
We should force client eval on Cast operator to mimic its type asserting functionality. However, some queries inject unnecessary casts, e.g.

```
from Customer c in ctx.Customers
select c.Name
```

we should strip casts for those cases (where the cast type is the same as element type of EntityQueryable

**Comment by smitpatel** [comment]
1) would be irrelevant now that we don't have client eval.

**Comment by smitpatel** [comment]
2nd case is fixed in 3.1 release.
